### DISCUSSION TITLE
thin films in MEEP and problems with resolution

### DISCUSSION THREAD

**Discussion by 10UAikawa** [NanoComp/meep#2720] [Q&A]
Hi MEEP community, a problem I am having is the inability to work with thin films in MEEP. So, for example if have dimensions that are in microns and the dimensions of parts of my structure is on the order of 400um and the smallest part of my design is 0.02um or 200nm. There is a large difference between the sizes. Thus, when I set my resolution the largest, I can set my resolution is 0.1 pixel/um before its starts to become too computationally expensive. With this resolution I am unable to distinguish features that are 0.02um in size. 

So how can this problem be reconciled? Are there features in MEEP that allow for adaptive meshing for objects? Are there any work arounds?

Thanks in Advance.

**Comment by smartalecH** [comment]
This is a FAQ: https://meep.readthedocs.io/en/latest/FAQ/#does-meep-support-grids-with-non-uniform-discretization

**Comment by stevengj** [comment]
When you have such a large separation of scales, it's better to handle some scale semi-analytically rather than trying to brute-force it. e.g. if you have a 400µm thick layer that is homogeneous, and wavelengths λ ≈ 1µm, you are probably better off doing a near-to-farfield transformation or similar to propagate light through the thick homogeneous layer.  But it's hard to give you better advice without knowing more details of your problem.

**Reply by 10UAikawa** [reply]
The wavelength of light I am concerned with is λ ≈ 750µm. This details my problem more broadly: https://github.com/NanoComp/meep/discussions/2718. But I was hoping to eventually use the adjoint method to optimize my design.

**Reply by Nikolaos-Matthaiakakis** [reply]
I often face this issue as I work with 2D materials that often have interesting optical properties in THz frequencies. It is true that semi-analytical methods can help but if more complex geometries or materials are required both in small and large dimensions it is still an issue. Until now I have managed to overcome the issue and even published some very nice publications thanks to meep (which is by far my favorite simulation tool) by limiting the structure dimensions as much as possible (for example https://doi.org/10.1002/adom.202303181) . On the other hand i have faced issues when trying to pursue structures with bigger dimensions and complexity due to the high resolution the graphene (or other 2D material) volumetric permitivity method demands.

1) In the case of 2D materials a solution would be to introduce a surface layer with a surface conductivity instead of volumetric geometries with volumetric permittivity. For example as recommended here https://olympias.lib.uoi.gr/jspui/bitstream/123456789/32331/1/PhD_thesis_Doukas.pdf in chapter 4.2.5. The volumetric method assumes a graphene thickness equal to the grid spacing so it can still be extremely demanding while the surface method does not suffer from this drawback and seems to be an elegant and efficient solution that might actually be easy to implement in meep (the chapter describes this in more detail). The surface conductivity method is currently implemented in some commercial tools. 

2) Of course a non-uniform grid would allow meep to be flexible in simulations that demand structures with large dimension mismatches even for non 2D surfaces. Overall this would be my dream update for meep as it would make almost all my research ideas much easier to implement (while currently some might be impossible to do with meep). Nevertheless I understand that unlike the suggestion in 1), a non-uniform grid is a very demanding effort.

I am curious if 1) were to be implemented if it could even be done by someone outside of the main meep development team, I do believe it wouldn't be too difficult to implement. On the other hand I imagine 2) would require very deep knowledge of the meep code.
